server/routerlicious/packages/services/src/mongodb.ts

```diff
@@ -7,103 +7,172 @@ import { assert } from "console";
 import * as core from "@fluidframework/server-services-core";
 import { AggregationCursor, Collection, MongoClient, MongoClientOptions } from "mongodb";
 import { Lumberjack } from "@fluidframework/server-services-telemetry";
+import { requestWithRetry } from "@fluidframework/server-services-core";
+import { MongoErrorRetryAnalyzer } from "./mongoExceptionRetryRules";
 
 const MaxFetchSize = 2000;
 
-export class MongoCollection<T> implements core.ICollection<T> {
-    constructor(private readonly collection: Collection<T>) {
+export class MongoCollection<T> implements core.ICollection<T>, core.IRetryAble {
+    constructor(
+        private readonly collection: Collection<T>,
+        public readonly retryEnabled = false,
+        private readonly telemetryEnabled = false,
+        private readonly mongoErrorRetryAnalyzer: MongoErrorRetryAnalyzer,
+    ) { }
+
+    public async aggregate(pipeline: any, options?: any): Promise<AggregationCursor<T>> {
+        const req = async () => new Promise<AggregationCursor<T>>(
+            (resolve) => resolve(this.collection.aggregate(pipeline, options)));
+        return this.requestWithRetry(req, "MongoCollection.aggregate");
     }
 
-    public aggregate(pipeline: any, options?: any): AggregationCursor<T> {
-        return this.collection.aggregate(pipeline, options);
-    }
-
-    // eslint-disable-next-line @typescript-eslint/ban-types,@typescript-eslint/promise-function-async
-    public find(query: object, sort: any, limit = MaxFetchSize, skip?: number): Promise<T[]> {
-        let queryCursor = this.collection
-            .find(query)
-            .sort(sort)
-            .limit(limit);
-
-        if (skip) {
-            queryCursor = queryCursor.skip(skip);
-        }
-        return queryCursor.toArray();
+    public async find(query: object, sort: any, limit = MaxFetchSize, skip?: number): Promise<T[]> {
+        const req: () => Promise<T[]> = async () => {
+            let queryCursor = this.collection
+                .find(query)
+                .sort(sort)
+                .limit(limit);
+
+            if (skip) {
+                queryCursor = queryCursor.skip(skip);
+            }
+            return queryCursor.toArray();
+        };
+        return this.requestWithRetry(req, "MongoCollection.find");
     }
 
-    // eslint-disable-next-line @typescript-eslint/ban-types,@typescript-eslint/promise-function-async
-    public findOne(query: object): Promise<T> {
-        return this.collection.findOne(query);
+    public async findOne(query: object): Promise<T> {
+        const req: () => Promise<T> = async () => this.collection.findOne(query);
+        return this.requestWithRetry(
+            req, // request
+            "MongoCollection.findAll", // callerName
```

Comment:
>findAll [](http://example.com/codeflow?start=29&length=7)

findOne?

Reply:
Nice find, addressed in d4556832bfd0a9c0cb5bde9bbacb337ab63974ee